### ISSUE TITLE
no_connection error connecting to Cumulus Linux switches

### ISSUE THREAD

**Issue by wmclendon** [ytti/oxidized#802]
Hello,

installed oxidized from Master on Ubuntu 16.04, I have net-ssh 3.0.2 installed with Ruby 2.3, and I get errors like the following when connecting to Cumulus Linux devices:

oxidized[2817]: W, [2017-04-07T10:27:07.374637 #2817]  WARN -- : 10.10.10.152 raised Timeout::Error with msg "execution expired"
oxidized[2817]: W, [2017-04-07T10:27:07.388513 #2817]  WARN -- : cumulus02 status no_connection, retry attempt 2

it eventually gives up and moves on.  per 'gem list' output I have oxidized 0.19.0

I tried downrevving to net-ssh 3.0.1 and 3.0.0, and even just doing 'gem install net-ssh' for the newest version (4.x I think), but either it didn't help, or oxidized wouldn't start, depending on the version I tried.

Anyone else having this issue?  If I ssh as the username / pass from the box I can login just fine.

Thanks,

Will

**Comment by wmclendon** [comment]
well I figured out the issue.  and there were actually 2 of them.  All of our Cumulus are running CL 3.2.1 at this time.  Once I figured out how to enable debug I could see where it was failing.  It was logging in and gathering data, and then bombing out on certain parts of files it was gathering because the prompt Regex wouldn't match.  

The first one was with /etc/quagga/Quagga.conf, where there was no blank line at the end of the file, so a 'cat /etc/quagga/Quagga.conf' returned something like this:

`..rest of quagga config...
vty
!user@switch:mgmt-vrf:~$`

which per Rubular did not match the defined prompt regex in the Cumulus.rb file of "/^((\w*)@(.*)):(.*)\$/" since \w does not match the !

I modified the regex to allow 0 or 1 ! at the beginning, and that got me past Quagga file, only to stumble again on the /etc/cumulus/ports file -- on our Dell S4048-ON 10G switches, this file again has no newline at the end of the file, so the prompt ends up like:

`53=40G
54=40Guser@switch:mgmt-vrf:~$`

which again the '=' is not matched by the \w bit of the regex so it bombed out there.

The Quagga newline issue is on my side I guess from provisioning; our template now has a newline at the end.  The cumulus ports file however, maybe that is something that changed from one version of Cumulus to another to not have the newline.

ultimately for now i've changed the prompt regex to be this ugly thing:
`prompt /^((!?)|(\w*=\w*))((\w*)@(.*)):/`

which basically allows it to match 0 or 1 '!' at beginning, or allow an '=' to be present in the prompt line as well.

Is it possible at all to have some improved error reporting where the syslog would report which command it could not complete due to waiting on prompt, or have it put in the backup something like 'unable to execute command XYZ due to waiting for prompt' rather than fail out the entire data backup?  Especially on Cumulus systems where the config is really many files, and not just 1 config file.


Thanks,

Will
